### ISSUE TITLE
New Algorithm: FSRS-7 + Logistic Regression Ensemble

### ISSUE THREAD

**Issue by HuyAiT** [open-spaced-repetition/srs-benchmark#331]
## FSRS-7 + Logistic Regression Ensemble

I opened a clean PR for this here: #334.

This is an experimental ensemble algorithm that combines FSRS-7 and Logistic Regression predictions in logit space. The motivation is to test whether a small linear feature-based model can complement FSRS-7 on some user histories while keeping the overall parameter count relatively low.

### Algorithm Overview

- **Base models**: FSRS-7 and Logistic Regression
- **Blending**: convert both predictions to logits, combine them as `w · logit_fsrs + (1 − w) · logit_lr`, then apply sigmoid
- **Separate blending weights** for same-day and non-same-day reviews
- **Weight fitting**: temporary FSRS-7 and LR models are trained on the first 70% of each user's training fold, then blending weights are fitted on out-of-sample predictions from the last 30% temporal holdout
- **Final inference**: after fitting the blending weights, both base models are retrained on the full per-user training fold before predicting the benchmark test fold
- **Regularization**: blending weights are regularized toward 0.4 and clamped to `[0.1, 0.9]`
- **Logged parameters**: 71 total (35 FSRS-7 + 34 LR + 2 blending weights)
- **Input features**: IL, FIL, G, SR

The out-of-sample temporal holdout is intended to avoid fitting the ensemble weights on predictions from models that already saw those same rows during training.

### Files

| File | Description |
|---|---|
| `models/fsrs7_lr_ensemble.py` | Ensemble model (new) |
| `features/fsrs7_lr_ensemble_engineer.py` | Feature engineer — builds both FSRS-7 tensors and LR features (new) |
| `evaluate.py` | Registered ensemble (71 params) |
| `config.py` | Added model name |
| `script.py` | Ensemble handling in training/inference loop |
| `models/__init__.py`, `models/model_factory.py`, `features/factory.py` | Registry entries |

### Local Validation

I ran a local validation on 500 sampled users from the 10k dataset, using seed = 2026 and explicitly including one previously problematic user (`241`). Both models were run on the same 500 users with:

```bash
python script.py --algo FSRS-7 --short --secs --equalize_test_with_non_secs
python script.py --algo FSRS-7-LR-Ensemble --short --secs --equalize_test_with_non_secs
```

Weighted by reviews on those 500 users:

| Metric | FSRS-7 | FSRS-7-LR-Ensemble | Delta |
|---|---:|---:|---:|
| LogLoss ↓ | 0.3062 | 0.3024 | -0.0038 |
| RMSE(bins) ↓ | 0.0466 | 0.0422 | -0.0044 |
| AUC ↑ | 0.7085 | 0.7158 | +0.0073 |
| MBE | 0.0034 | 0.0037 | +0.0003 |

Weighted by users on the same sample:

| Metric | FSRS-7 | FSRS-7-LR-Ensemble | Delta |
|---|---:|---:|---:|
| LogLoss ↓ | 0.3406 | 0.3359 | -0.0047 |
| RMSE(bins) ↓ | 0.0649 | 0.0604 | -0.0045 |
| AUC ↑ | 0.7048 | 0.7119 | +0.0071 |
| MBE | 0.0035 | 0.0035 | +0.0000 |

Per-user wins on the 500-user sample:

- LogLoss: 412 / 500
- RMSE(bins): 397 / 500
- AUC: 380 / 500
- absolute MBE: 342 / 500

Important caveat: this is still only a sampled local benchmark, not the full 10k-user benchmark. The ensemble also still underperforms FSRS-7 on some individual users. For example, user `241` remains worse for the ensemble, especially on AUC, so I do not want to overstate the result. The sample suggests the method may help on average, especially for LogLoss/RMSE, but the full benchmark is needed before drawing a strong conclusion.

### How to run

```bash
python script.py --algo FSRS-7-LR-Ensemble --short --secs --equalize_test_with_non_secs
```

If you think this model is worth evaluating, would it be possible to run the full 10k-user benchmark on your side? If not, I can continue running it locally, though it may take quite a while on my machine.

I'm very happy to revise or simplify the implementation if this approach does not fit the benchmark design, or if you would prefer a different validation protocol for ensemble/stacking models. Please feel free to point out anything that looks off.
